src/Http/Http.Abstractions/src/Internal/PathStringHelper.cs

```diff
@@ -24,24 +27,57 @@ internal class PathStringHelper
             true,  true,  true,  false, false, false, true,  false,     // 0x78 - 0x7F
         };
 
+        // Map from an ASCII char to its hex value, e.g. arr['b'] == 11. 0xFF means it's not a hex digit.
+        // From coreclr/src/System.Private.CoreLib/shared/System/Number.Parsing.cs
+        // byte[] rather than ReadOnlySpan<byte> to avoid double .Length check https://github.com/dotnet/coreclr/issues/24209
+        private static readonly byte[] CharToHexLookup = new byte[]
+        {
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 15
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 31
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 47
+            0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  0x8,  0x9,  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 63
+            0xFF, 0xA,  0xB,  0xC,  0xD,  0xE,  0xF,  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 79
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 95
+            0xFF, 0xa,  0xb,  0xc,  0xd,  0xe,  0xf // 102
+        };
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsValidPathChar(char c)
         {
-            return c < ValidPathChars.Length && ValidPathChars[c];
+            // Use local array and uint .Length compare to elide the bounds check on array access
+            var validChars = ValidPathChars;
+            var i = (int)c;
+            return (uint)i < (uint)validChars.Length && validChars[i];
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsPercentEncodedChar(string str, int index)
         {
-            return index < str.Length - 2
-                && str[index] == '%'
-                && IsHexadecimalChar(str[index + 1])
-                && IsHexadecimalChar(str[index + 2]);
+            if (str[index] == '%' && index < str.Length - 2)
+            {
+                return AreFollowingTwoCharsHex(str, index);
+            }
+
+            return false;
         }
 
-        public static bool IsHexadecimalChar(char c)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static bool AreFollowingTwoCharsHex(string str, int index)
         {
-            return ('0' <= c && c <= '9')
-                || ('A' <= c && c <= 'F')
-                || ('a' <= c && c <= 'f');
+            Debug.Assert(index < str.Length - 2);
+
+            var charToHex = CharToHexLookup;
+            var i1 = (int)str[index + 1];
+            var i2 = (int)str[index + 2];
+            if ((uint)i1 < (uint)charToHex.Length &&
+                charToHex[i1] != 0xff &&
+               (uint)i2 < (uint)charToHex.Length &&
+                charToHex[i2] != 0xff)
+            {
+                return true;
```

Comment:
Maybe the following is better?
```c#
if ((uint)i1 < (uint)charToHex.Length
 && (uint)i2 < (uint)charToHex.Length)
{
    var tmp = charToHex[i1] | charToHex[i2];
    return tmp != 0xFF;
}

return false;
```

```diff
       3BF2                 cmp      esi, edx
       7624                 jbe      SHORT G_M42979_IG04
-      4863D2               movsxd   rdx, edx
-      807C1010FF           cmp      byte  ptr [rax+rdx+16], 255
-      7415                 je       SHORT G_M42979_IG04
-      3BF7                 cmp      esi, edi
-      7611                 jbe      SHORT G_M42979_IG04
-      4863FF               movsxd   rdi, edi
-      807C3810FF           cmp      byte  ptr [rax+rdi+16], 255
-      7407                 je       SHORT G_M42979_IG04
-      B801000000           mov      eax, 1
+      3BF7                 cmp      esi, edi
+      7620                 jbe      SHORT G_M42979_IG04
+      4863F2               movsxd   rsi, edx
+      400FB6743010         movzx    rsi, byte  ptr [rax+rsi+16]
+      4863FF               movsxd   rdi, edi
+      0FB6443810           movzx    rax, byte  ptr [rax+rdi+16]
+      0BC6                 or       eax, esi
+      3DFF000000           cmp      eax, 255
+      0F95C0               setne    al
+      0FB6C0               movzx    rax, al

G_M42979_IG03:
       5D                   pop      rbp
       C3                   ret 
```

Safes two branches.

Reply:
Done